aggregator/entry.go

```diff
@@ -315,170 +339,179 @@ func (e *Entry) recordLastAccessed(currTime time.Time) {
 		}
 	}
 
-	// We should update the policies first.
-	newAggregations := make(map[policy.Policy]*list.Element, len(policies))
-	for _, p := range policies {
-		if elem, exists := e.aggregations[p]; exists {
-			newAggregations[p] = elem
-		} else {
-			aggTypes, err := e.decompressor.Decompress(p.AggregationID)
-			if err != nil {
-				return err
-			}
+	// Otherwise it is necessary to make a copy because it's not owned by us.
+	elemID := make(metricid.RawID, len(metric.ID))
+	copy(elemID, metric.ID)
+	return elemID
+}
+
+func (e *Entry) updateMetadatasWithLock(
+	metric unaggregated.MetricUnion,
+	hasDefaultMetadatas bool,
+	sm metadata.StagedMetadata,
+) error {
+	var (
+		elemID          = e.maybeCopyIDWithLock(metric)
+		newAggregations = make(aggregationValues, 0, initialAggregationCapacity)
+	)
 
-			var newElem metricElem
-			switch typ {
-			case unaggregated.CounterType:
-				if !aggTypes.IsValidForCounter() {
-					return fmt.Errorf("invalid aggregation types %s for Counter", aggTypes.String())
+	// Update the metadatas.
+	for _, pipeline := range sm.Pipelines {
+		storagePolicies := e.storagePolicies(pipeline.StoragePolicies)
+		for _, storagePolicy := range storagePolicies {
+			key := aggregationKey{
+				aggregationID: pipeline.AggregationID,
+				storagePolicy: storagePolicy,
+				pipeline:      pipeline.Pipeline,
+			}
+			// Remove duplicate aggregation pipelines.
+			if newAggregations.contains(key) {
+				continue
+			}
+			if idx := e.aggregations.index(key); idx >= 0 {
+				newAggregations = append(newAggregations, e.aggregations[idx])
+			} else {
+				aggTypes, err := e.decompressor.Decompress(key.aggregationID)
+				if err != nil {
+					return err
 				}
-				newElem = e.opts.CounterElemPool().Get()
-			case unaggregated.BatchTimerType:
-				if !aggTypes.IsValidForTimer() {
-					return fmt.Errorf("invalid aggregation types %s for Timer", aggTypes.String())
+				var newElem metricElem
+				switch metric.Type {
+				case unaggregated.CounterType:
+					if !aggTypes.IsValidForCounter() {
+						return fmt.Errorf("invalid aggregation types %s for Counter", aggTypes.String())
+					}
+					newElem = e.opts.CounterElemPool().Get()
+				case unaggregated.BatchTimerType:
+					if !aggTypes.IsValidForTimer() {
+						return fmt.Errorf("invalid aggregation types %s for Timer", aggTypes.String())
+					}
+					newElem = e.opts.TimerElemPool().Get()
+				case unaggregated.GaugeType:
+					if !aggTypes.IsValidForGauge() {
+						return fmt.Errorf("invalid aggregation types %s for Gauge", aggTypes.String())
+					}
+					newElem = e.opts.GaugeElemPool().Get()
+				default:
+					return errInvalidMetricType
 				}
-				newElem = e.opts.TimerElemPool().Get()
-			case unaggregated.GaugeType:
-				if !aggTypes.IsValidForGauge() {
-					return fmt.Errorf("invalid aggregation types %s for Gauge", aggTypes.String())
+				// NB: The pipeline may not be owned by us and as such we need to make a copy here.
```

Comment:
who else might own the pipeline?

Reply:
So the thinking here is that the pipeline might be reused by the decoding iterator, just like it does with metric IDs, and as such we need to make a copy to ensure the pipeline doesn't get changed underneath us.